### ISSUE TITLE
--max-jobs 0 is ignored for preferLocalBuild derivations

### ISSUE THREAD

**Issue by catern** [NixOS/nix#1798]
The following command:

```
nix-build -j0 --expr 'derivation { system = builtins.currentSystem; name = "bad"; builder = "false"; preferLocalBuild = true; }'
```

should not build; rather it should fail with: "error: unable to start any build; either increase ‘--max-jobs’ or enable distributed builds".

Yet it does (try to) build:
```
catern@sandbox ~ $ nix-build -j0 --expr 'derivation { system = builtins.currentSystem; name = "bad"; builder = "false"; preferLocalBuild = true; }'
these derivations will be built:
  /nix/store/w4smiql512izr7859w62r8dbkp1hvhi1-bad.drv
building path(s) ‘/nix/store/jskp2i1jvnhhgq1ky5vc889ksgai69kk-bad’
while setting up the build environment: executing ‘false’: No such file or directory
builder for ‘/nix/store/w4smiql512izr7859w62r8dbkp1hvhi1-bad.drv’ failed with exit code 1
error: build of ‘/nix/store/w4smiql512izr7859w62r8dbkp1hvhi1-bad.drv’ failed
```

Among other things, this makes it harder to work around the lack of https://github.com/NixOS/nix/issues/1734 because builds will still be carried out when -j0 is passed, rather than forcing either substitution or immediate failure.

**Comment by catern** [comment]
I believe this is the culprit:

https://github.com/NixOS/nix/blob/master/src/libstore/build.cc#L1440
```
    /* Make sure that we are allowed to start a build.  If this
       derivation prefers to be done locally, do it even if
       maxBuildJobs is 0. */
    unsigned int curBuilds = worker.getNrLocalBuilds();
    if (curBuilds >= settings.maxBuildJobs && !(buildLocally && curBuilds == 0)) {
```

buildLocally is set for preferLocalBuild derivations.

Is there a reason that local builds should happen even if maxBuildJobs is 0? Can we just remove this escape hatch? I think if the user sets maxBuildJobs to 0, we really should not start any jobs.

**Comment by catern** [comment]
I'm using the following workaround:

```
nix-instantiate --read-write-mode --eval --expr '__storePath [insert_store_path_here]'
```

This is essentially a hacky way to call ensurePath for that path. ensurePath fails immediately if the path is not valid and cannot be substituted, without trying to build. That's exactly what I want, so this workaround is perfectly functional for me. You can see how I use it in: https://github.com/catern/nix-utils/blob/master/fixedout.py#L53

Nevertheless, I think the original bug here should be fixed. Also, maybe a direct interface should be provided to use ensurePath?

**Comment by offlinehacker** [comment]
This bug pretty annoying especially if you have slow upload, as it first downloads source files locally, and then uploads them to target machine, kinda does not make sense.

**Comment by catern** [comment]
I still care about this.

**Comment by catern** [comment]
I still care about this.

**Comment by maralorn** [comment]
Good news, @catern: `--max-jobs 0` is respected for local builds in nix 2.4.

**Comment by catern** [comment]
Nice! For me, anyway. I see in #5646 that you don't like this change. My ultimate desire is a way to do #1734 so if we have to give up on a fix for #1798, that would be fine if we can make some progress on #1734.

**Comment by tomberek** [comment]
This looks resolved in post 2.4, though there is an ongoing discussion if this should be the behavior or not. (see linked issues)
